docs/arch/adr-18.md

```diff
@@ -0,0 +1,47 @@
+# ADR 18: Workflow routing
+
+Created: September 4, 2019
+
+## Context
+
+On PFE, there are a couple of different ways workflows are routed:
+
+- Navigating to `/classify`, the user sees either the default workflow for that project, a random active workflow, or the workflow stored in project preferences for logged-in volunteers.
+- Navigating to `/classify?workflow=[workflow_id]` shows the workflow determined by the query parameter when the "Use workflow query" feature is enabled for the project, or if the user has the correct role (owner, collaborator, tester, or admin).
+
+This is not optimal for several reasons. Users cannot bookmark specific workflows unless they use the query parameter; there can be uncertainty over which actual workflow is being shown (particularly as the workflow name isn't shown in the UI); different workflows can be shown at the same URL at different times; and ultimately, workflows are static resources that should be routed to - that's what URLs are for.
+
+Additionally, [we currently have a requirement for CSSI CitSci to be able to route to specific subjects](https://github.com/zooniverse/front-end-monorepo/issues/806#issuecomment-495685027). It's conceivable that a project will have a requirement for routing to a specific subject set linked to a workflow as well.
+
+## Decision
+
+We adopt the following URL structure:
+
+```
+/projects/[owner]/[project]/classify/workflow/:workflow-id/subject-set/:subject-set-id/subject/:subject-id
+```
+
+This would be facilitated by the [dynamic routing feature in Next.js 9](https://github.com/zeit/next.js/#dynamic-routing) (see [#1071](https://github.com/zooniverse/front-end-monorepo/pull/1071)).
+
+Workflow links from the home page would route directly to their workflow-specific URLs.
+
+### `/classify` behaviour
+
+When navigating to `/classify`, the page should redirect the user to the correct workflow URL in the following priority:
+
+1. Workflow ID set in user project preferences by the user
+1. Workflow ID set in user project preferences by the project
+1. Project default workflow
+1. Random active selection
+
+The new URL that we redirect the user to should replace the entry for `/classify` in the browser history.
```

Comment:
Query: how descriptive will the new URL be?

For example, let's say we have Project 'A Space Odyssey' with Workflow 2001 (default).

If I type in `zooniverse.org/projects/kubrick/a-space-odyssey/classify`, then (assuming no project preferences) I'd be redirected to/the browser history will rewrite itself to `zooniverse.org/projects/kubrick/a-space-odyssey/classify/2001` - correct?

I just wanted to check that it's this straightforward, because given the above statement...

> ... we currently have a requirement for CSSI CitSci to be able to route to specific subjects ...
> ...
> We adopt the following URL structure:
> `/projects/[owner]/[project]/classify/workflow/:workflow-id/subject-set/:subject-set-id/subject/:subject-id`

... I wasn't entirely sure if `/classify` would instead redirect me to `zooniverse.org/projects/kubrick/a-space-odyssey/classify/2001/subject-set/3000/subject/44444`, with the URL changing every time I'm served a new Subject. (Because that'd make bookmarking my current workflow a bit tricky.)

Reply:
I'm assuming (but @srallen should confirm) that per-subject URLs won't be enabled for every project. Otherwise, yes, workflow bookmarking would be of less use - I guess you could pick up from that subject, and you'd get a banner saying you've already classified it and asking you to move on.

Another thing I've just thought of is where on the URL chain we handle redirects. We could cascade them: `classify` could redirect the user to `workflow`, which could _then_ redirect to `subject_set` etc. It would keep the code simple, but result in multiple redirects which would be a sub-par user experience. 

Alternatively, `classify` could bump you straight off to `/classify/2001/subject-set/3000/subject/44444` - more complex, but only one redirect.

This is an implementation issue that's outside the ADR, but worth thinking about

Reply:
The initial implementation will only be for the citsci.org use case. They will generate a link in the UI of their app that the volunteer can click and load a specific subject to classify on. I don't think we need to keep other projects from being able to load specific subject sets and subjects to classify on, but we won't be advertising or building a UI for it on our site (maybe eventually, but this won't be part of CSSI). For this reason, I don't think we should generally be redirecting all projects to specific subject classification URLs as we don't know what the UI and potential side effects will be. I think the only redirect we support should be for `/classify/workflows/:workflow-id`

As for what the URL behavior should be for the citsci.org case as the volunteer is classifying? I think we can address that (with another additional ADR if necessary) when we start to build out that functionality.